### DISCUSSION TITLE
MCS, R-group Decomposition (RGD) and Isotopes

### DISCUSSION THREAD

**Discussion by chloechow** [rdkit/rdkit#6495] [Q&A]
Hello RDKit team,

I've been trying to find the maximum common substructure (MCS) and then perform  R-group decomposition (RGD) on a set of molecules with isotopes using RDKit, but I'm encountering some issues.

Here are the steps I'm trying to do:  
```
mols = [Chem.MolFromSmiles(m) for m in ['n1c(nc(cc1)c1cn(c2c1cccc2)C)Nc1c(cc(c(c1)NC(=O)C=C)N(CCN(C)C)C)OC', 'n1c(nc(cc1)c1cn(c2c1cccc2)C)Nc1c(cc(c(c1)NC(=O)C(=C([2H])[2H])[2H])N(CCN(C)C)C)OC', 'n1c(nc(cc1)c1cn(c2c1cccc2)C([2H])([2H])[2H])Nc1c(cc(c(c1)NC(=O)C(=C([2H])[2H])[2H])N(CCN(C)C)C)OC',  'n1c(nc(cc1)c1cn(c2c1cccc2)C([2H])([2H])[2H])Nc1c(cc(c(c1)NC(=O)C=C)N(C(C(N(C([2H])([2H])[2H])C([2H])([2H])[2H])([2H])[2H])([2H])[2H])C([2H])([2H])[2H])OC', 'n1c(nc(cc1)c1cn(c2c1cccc2)C([2H])([2H])[2H])Nc1c(cc(c(c1)NC(=O)C(=C([2H])[2H])[2H])N(C(C(N(C([2H])([2H])[2H])C([2H])([2H])[2H])([2H])[2H])([2H])[2H])C([2H])([2H])[2H])OC']]
mcs_params = rdFMCS.MCSParameters()
mcs_params.AtomCompareParameters.matchIsotope = True
mcs = rdFMCS.FindMCS(mols, mcs_params)
core = Chem.MolFromSmarts(mcs.smartsString)

ps = rdRGroupDecomposition.RGroupDecompositionParameters()
ps.RGroupLabels = rdRGroupDecomposition.RGroupLabels.IsotopeLabels
ps.RGroupLabelling = rdRGroupDecomposition.RGroupLabelling.Isotope
groups, _ = rdRGroupDecomposition.RGroupDecompose([core],mols,asRows=False,options=ps)

print(groups.keys())
```
The output:
```
dict_keys(['Core'])
```

The isotopes are not being decomposed correctly in the RGD process. Could you please provide some guidance on how to handle isotopes in R-group decomposition with RDKit? Any help would be greatly appreciated.

Thank you in advance for your assistance.

Configuration:
- RDKit Version: 2023.03.1
- Operating system: CentOS Linux release 8.5.2111
- Python version: 3.11.3
- Are you using conda? Yes
- If you are using conda, which channel did you install the rdkit from? conda-forge

**Comment by bp-kelley** [comment]
The isotope labels mentioned in the parameters correspond to the ones in
the core.  I.e. an isotope in the core labeled [*1] will become rgroup 1.
Here you have no such labelling in the core so these labels will never be
used:

'[#7]1:[#6](:[#7]:[#6](:[#6]:[#6]:1)-[#6]1:[#6]:[#7](:[#6]2:[#6]:1:[#6]:[#6]:[#6]:[#6]:2)-[#6])-[#7]-[#6]1:[#6](:[#6]:[#6](:[#6](:[#6]:1)-[#7]-[#6](=[#8])-[#6]=[#6])-[#7](-[#6]-[#6]-[#7](-[#6])-[#6])-[#6])-[#8]-[#6]'


What is your intended output here?

**Reply by chloechow** [reply]
Thank you for responding. I believe I may have misunderstood the RGD parameters. My intention is to treat each unique isotope as an individual R-group while maintaining the core without any deuterium substitution. Could you please advise me if my method of finding the MCS was incorrect? Your guidance would be greatly appreciated. Thank you.

**Reply by chloechow** [reply]
Here is my intended output.
![image](https://github.com/rdkit/rdkit/assets/50295474/5bd108b8-68e0-435b-8a24-41098b97f5b3)

**Comment by chloechow** [comment]
Despite my attempts to directly define the core, I was unable to detect any R-groups.
```
mols = [Chem.MolFromSmiles(m) for m in ['n1c(nc(cc1)c1cn(c2c1cccc2)C)Nc1c(cc(c(c1)NC(=O)C=C)N(CCN(C)C)C)OC', 'n1c(nc(cc1)c1cn(c2c1cccc2)C)Nc1c(cc(c(c1)NC(=O)C(=C([2H])[2H])[2H])N(CCN(C)C)C)OC', 'n1c(nc(cc1)c1cn(c2c1cccc2)C([2H])([2H])[2H])Nc1c(cc(c(c1)NC(=O)C(=C([2H])[2H])[2H])N(CCN(C)C)C)OC',  'n1c(nc(cc1)c1cn(c2c1cccc2)C([2H])([2H])[2H])Nc1c(cc(c(c1)NC(=O)C=C)N(C(C(N(C([2H])([2H])[2H])C([2H])([2H])[2H])([2H])[2H])([2H])[2H])C([2H])([2H])[2H])OC', 'n1c(nc(cc1)c1cn(c2c1cccc2)C([2H])([2H])[2H])Nc1c(cc(c(c1)NC(=O)C(=C([2H])[2H])[2H])N(C(C(N(C([2H])([2H])[2H])C([2H])([2H])[2H])([2H])[2H])([2H])[2H])C([2H])([2H])[2H])OC']]
core = Chem.MolFromSmarts('COc(c(Nc1nc(-c2cn(C[*:8])c3c2cccc3)ccn1)c1)cc(N(C[*:4])C(C(N(C[*:6])C[*:7])[*:5])[*:3])c1NC(C([*:1])=C[*:2])=O')

ps = rdRGroupDecomposition.RGroupDecompositionParameters()
groups, _ = rdRGroupDecomposition.RGroupDecompose([core],mols,asRows=False,options=ps)

print(groups.keys())
```
The output:
```
dict_keys(['Core'])
```

**Reply by bp-kelley** [reply]
Hydrogens are odd ducks here, they are treated differently than other atom types for good reasons, but they do make this use-case difficult.  I'm on the fence whether I consider this a bug or not.

**Reply by bp-kelley** [reply]
However, you'll note that you have four different cores, each with a different hydrogen pattern.  Here is a simpler test case:

```
mols = [Chem.MolFromSmiles(s) for s in ['c1ccccc1', 'c1ccccc1[2H]']]
core = Chem.MolFromSmarts("c1ccccc1")
params = rgd.RGroupDecompositionParameters()
params.removeAllHydrogenRGroups = False
rgd.RGroupDecompose(core, mols, asSmiles=True, options=params)
```

```
([{'Core': 'c1ccccc1'}, {'Core': '[2H]c1ccccc1'}], [])
```

Certainly not what you want here...
